bitbybit/src/bitenum.rs

```diff
@@ -1,303 +1,225 @@
-                }
-            } else {
-                quote! {
-                    #( #cfg_attributes )*
-                    #expression => Self::#name,
-                }
-            }
-        })
-        .collect();
+    let bits = config.bits;
+    let (base_type, qualified_type) = (bits.base_type()?, bits.qualified_path()?);
+    let (constructor, reader) = (bits.constructor()?, bits.reader());
 
-    let constructor_function = if return_is_result {
-        quote!(
-            /// Creates a new instance of this bitfield with the given raw value, or
-            /// Err(value) if the value does not exist in the enum.
-            pub const fn new_with_raw_value(value: #bounded_data_type) -> Result<Self, #base_data_type> {
-                match value #bounded_getter {
-                    #( #case_values )*
-                    _ => Err(value #bounded_getter)
-                }
-            }
-        )
-    } else {
-        let panic_string = format!("{}: Unhandled value", enum_name);
-        quote!(
-            /// Creates a new instance of this bitfield with the given raw value.
-            ///
-            /// As the enum is exhaustive, this function will always return a valid result
-            pub const fn new_with_raw_value(value: #bounded_data_type) -> Self {
-                match value #bounded_getter {
-                    #( #case_values )*
-                    _ => panic!(#panic_string)
-                }
-            }
-        )
+    let non_exhaustive = config.exhaustive.matches(false);
+    let ok = non_exhaustive.then_some(quote!(Ok));
+    let new_return_type = match non_exhaustive {
+        true => quote!(std::result::Result<Self, #base_type>),
+        false => quote!(Self),
     };
-
-    let expanded = quote! {
+    let new_default_branch = match non_exhaustive {
+        true => quote!(value => Err(value)),
+        false => quote!(_ => unreachable!()),
+    };
+    let new_match_branches = input.variants.iter().map(|variant| {
+        let cfg_attrs = variant.attrs.iter().filter(|a| conditional_attr(a));
+        let value = &variant.discriminant.as_ref().unwrap().1;
+        let variant_name = &variant.ident;
+        quote!( #( #cfg_attrs )* (#value) => #ok(Self::#variant_name) )
+    });
+    let (attrs, vis, name, variants) = (&input.attrs, &input.vis, &input.ident, &input.variants);
+    Ok(quote! {
         #[derive(Copy, Clone)]
-        #( #enum_attrs )*
-        #enum_vis enum #enum_name {
+        #( #attrs )*
+        #vis enum #name {
             #variants
         }
 
-        impl #enum_name {
-            /// Returns the underlying raw value of this bitfield
-            pub const fn raw_value(self) -> #bounded_data_type { #result_constructor(self as #base_data_type) }
+        impl #name {
+            /// Returns the underlying raw value of this bitfield.
+            pub const fn raw_value(self) -> #qualified_type {
+                #constructor(self as #base_type)
```

Comment:
The name 'constructor' is a bit misleading. Without context I'd expect a constructor for Self here (pretty much what new_with_raw_value() is).

Maybe call it raw_value_constructor?